### ISSUE TITLE
Observing errors on analytics console when EI publish data to analytics

### ISSUE THREAD

**Issue by gvasanka** [wso2/product-ei#457]
Steps:
1. Configure EI to publish data to Analytics as mention in following doc
https://docs.wso2.com/display/EI610/Prerequisites+to+Publish+Statistics
2. Start the Analytics Server
3. Start the Integrator server
4. Invoke the statics enabled proxy service
5. Check the analytics stack trace

Note: Analytics and Integrator run on IBM JDK 1.8

Issue:

EI Analytics throwing following error when EI Publish data to Analytics

```
[2017-03-31 16:50:16,598] [EI-Analytics] ERROR {org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer} -  Error in processing index batch operations: this IndexWriter is closed
org.apache.lucene.store.AlreadyClosedException: this IndexWriter is closed
	at org.apache.lucene.index.IndexWriter.ensureOpen(IndexWriter.java:718)
	at org.apache.lucene.index.IndexWriter.ensureOpen(IndexWriter.java:732)
	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1359)
	at org.apache.lucene.index.IndexWriter.addDocument(IndexWriter.java:1142)
	at org.apache.lucene.facet.taxonomy.directory.DirectoryTaxonomyWriter.addCategoryDocument(DirectoryTaxonomyWriter.java:497)
	at org.apache.lucene.facet.taxonomy.directory.DirectoryTaxonomyWriter.internalAddCategory(DirectoryTaxonomyWriter.java:458)
	at org.apache.lucene.facet.taxonomy.directory.DirectoryTaxonomyWriter.addCategory(DirectoryTaxonomyWriter.java:425)
	at org.apache.lucene.facet.FacetsConfig.processFacetFields(FacetsConfig.java:324)
	at org.apache.lucene.facet.FacetsConfig.build(FacetsConfig.java:287)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.generateIndexDoc(AnalyticsDataIndexer.java:1542)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.updateIndex(AnalyticsDataIndexer.java:1416)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.processIndexOperationBatch(AnalyticsDataIndexer.java:349)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.processLocalShardDataQueue(AnalyticsDataIndexer.java:304)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.processIndexOperationsSlice(AnalyticsDataIndexer.java:235)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.processIndexOperations(AnalyticsDataIndexer.java:223)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.access$100(AnalyticsDataIndexer.java:85)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer$IndexWorker.run(AnalyticsDataIndexer.java:2184)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1153)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:785)
Caused by: java.nio.file.FileSystemException: /home/asankav/servers/releasetest/ei/alpha/wso2ei-6.1.0-ALPHA/wso2/analytics/repository/data/index_data/2/_data/taxonomy/-1234_org_wso2_esb_analytics_stream_mediatorstatpersecond/_j.fdx: Too many open files
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:103)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:114)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:119)
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:226)
	at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:445)
	at java.nio.file.Files.newOutputStream(Files.java:227)
	at org.apache.lucene.store.FSDirectory$FSIndexOutput.<init>(FSDirectory.java:262)
	at org.apache.lucene.store.FSDirectory.createOutput(FSDirectory.java:211)
	at org.apache.lucene.store.TrackingDirectoryWrapper.createOutput(TrackingDirectoryWrapper.java:43)
	at org.apache.lucene.codecs.compressing.CompressingStoredFieldsWriter.<init>(CompressingStoredFieldsWriter.java:110)
	at org.apache.lucene.codecs.compressing.CompressingStoredFieldsFormat.fieldsWriter(CompressingStoredFieldsFormat.java:128)
	at org.apache.lucene.codecs.lucene50.Lucene50StoredFieldsFormat.fieldsWriter(Lucene50StoredFieldsFormat.java:183)
	at org.apache.lucene.index.DefaultIndexingChain.initStoredFieldsWriter(DefaultIndexingChain.java:81)
	at org.apache.lucene.index.DefaultIndexingChain.startStoredFields(DefaultIndexingChain.java:258)
	at org.apache.lucene.index.DefaultIndexingChain.processDocument(DefaultIndexingChain.java:295)
	at org.apache.lucene.index.DocumentsWriterPerThread.updateDocument(DocumentsWriterPerThread.java:232)
	at org.apache.lucene.index.DocumentsWriter.updateDocument(DocumentsWriter.java:458)
	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1363)
	... 17 more
[2017-03-31 16:50:18,099] [EI-Analytics] ERROR {org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer} -  Error in processing index batch operations: this IndexWriter is closed
org.apache.lucene.store.AlreadyClosedException: this IndexWriter is closed
	at org.apache.lucene.index.IndexWriter.ensureOpen(IndexWriter.java:718)
	at org.apache.lucene.index.IndexWriter.ensureOpen(IndexWriter.java:732)
	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1359)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.updateIndex(AnalyticsDataIndexer.java:1416)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.processIndexOperationBatch(AnalyticsDataIndexer.java:349)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.processLocalShardDataQueue(AnalyticsDataIndexer.java:304)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.processIndexOperationsSlice(AnalyticsDataIndexer.java:235)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.processIndexOperations(AnalyticsDataIndexer.java:223)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer.access$100(AnalyticsDataIndexer.java:85)
	at org.wso2.carbon.analytics.dataservice.core.indexing.AnalyticsDataIndexer$IndexWorker.run(AnalyticsDataIndexer.java:2184)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1153)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:785)
Caused by: java.nio.file.FileSystemException: /home/asankav/servers/releasetest/ei/alpha/wso2ei-6.1.0-ALPHA/wso2/analytics/repository/data/index_data/3/_data/index/-1234_org_wso2_esb_analytics_stream_mediatorstatpersecond/_j.fdt: Too many open files
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:103)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:114)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:119)
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:226)
	at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:445)
	at java.nio.file.Files.newOutputStream(Files.java:227)
	at org.apache.lucene.store.FSDirectory$FSIndexOutput.<init>(FSDirectory.java:262)
	at org.apache.lucene.store.FSDirectory.createOutput(FSDirectory.java:211)
	at org.apache.lucene.store.TrackingDirectoryWrapper.createOutput(TrackingDirectoryWrapper.java:43)
	at org.apache.lucene.codecs.compressing.CompressingStoredFieldsWriter.<init>(CompressingStoredFieldsWriter.java:113)
	at org.apache.lucene.codecs.compressing.CompressingStoredFieldsFormat.fieldsWriter(CompressingStoredFieldsFormat.java:128)
	at org.apache.lucene.codecs.lucene50.Lucene50StoredFieldsFormat.fieldsWriter(Lucene50StoredFieldsFormat.java:183)
	at org.apache.lucene.index.DefaultIndexingChain.initStoredFieldsWriter(DefaultIndexingChain.java:81)
	at org.apache.lucene.index.DefaultIndexingChain.startStoredFields(DefaultIndexingChain.java:258)
	at org.apache.lucene.index.DefaultIndexingChain.processDocument(DefaultIndexingChain.java:295)
	at org.apache.lucene.index.DocumentsWriterPerThread.updateDocument(DocumentsWriterPerThread.java:232)
	at org.apache.lucene.index.DocumentsWriter.updateDocument(DocumentsWriter.java:458)
	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1363)
	... 10 more

```

**Comment by isurusuranga** [comment]
To avoid this issue please increase the open files limit of your Operating system. 'ulimit -n' would give you the current number (by default it is 1024 in most of the cases). You can uncomment
#ulimit -n 100000  (after uncommenting it would be : ulimit -n 100000 ) in the wso2server.sh file in order to increase the value. Could you please check and verify this ?

**Comment by gvasanka** [comment]
After setting ulimit -n 100000 also I am getting the too many open files issue. Please note I am running analytics on IBM JDK 1.8

Please check below stack trace

```
[2017-03-31 18:44:00,294] [EI-Analytics] ERROR {org.wso2.carbon.analytics.spark.core.AnalyticsTask} -  Error while executing the scheduled task for the script: EsbAnalytics-SparkScript-Realtime-Statistic
org.wso2.carbon.analytics.spark.core.exception.AnalyticsExecutionException: Exception in executing query INSERT INTO TABLE esbServiceStatPerHour        SELECT year, month, day, hour, componentId, componentName, componentType,        sum(totalDuration) as totalDuration, min(minDuration) as minDuration, max(maxDuration) as maxDuration,        sum(noOfInvocation) as noOfInvocation, sum(faultCount) as faultCount,        getHourStartingTime(year, month, day, hour) as startTime,        cast(getHourStartingTime(year, month, day, hour) as STRING) as facetStartTime, entryPoint, meta_tenantId,        getHourStartingTime(year, month, day, hour) as _timestamp        FROM esbServiceStatPerMinute        GROUP BY        year, month, day, hour, componentId, componentName, componentType, entryPoint, meta_tenantId
	at org.wso2.carbon.analytics.spark.core.internal.SparkAnalyticsExecutor.executeQueryLocal(SparkAnalyticsExecutor.java:772)
	at org.wso2.carbon.analytics.spark.core.internal.SparkAnalyticsExecutor.executeQuery(SparkAnalyticsExecutor.java:729)
	at org.wso2.carbon.analytics.spark.core.CarbonAnalyticsProcessorService.executeQuery(CarbonAnalyticsProcessorService.java:201)
	at org.wso2.carbon.analytics.spark.core.CarbonAnalyticsProcessorService.executeScript(CarbonAnalyticsProcessorService.java:151)
	at org.wso2.carbon.analytics.spark.core.AnalyticsTask.execute(AnalyticsTask.java:60)
	at org.wso2.carbon.ntask.core.impl.TaskQuartzJobAdapter.execute(TaskQuartzJobAdapter.java:67)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:522)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1153)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:785)
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 160.0 failed 1 times, most recent failure: Lost task 0.0 in stage 160.0 (TID 7323, localhost): java.io.FileNotFoundException: /home/asankav/servers/releasetest/ei/alpha/wso2ei-6.1.0-ALPHA/wso2/analytics/tmp/blockmgr-3ba41862-2602-498d-aeb9-50a9918d459f/11/temp_shuffle_c7b89fd2-8842-4421-9ac2-6bd93a656537 (Too many open files)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:286)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:226)
	at org.apache.spark.storage.DiskBlockObjectWriter.open(DiskBlockObjectWriter.scala:88)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:140)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:73)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:41)
	at org.apache.spark.scheduler.Task.run(Task.scala:89)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:227)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1153)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:785)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1431)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1419)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1418)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1418)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:799)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:799)
	at scala.Option.foreach(Option.scala:236)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:799)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1640)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1599)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1588)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
	at java.lang.Thread.getStackTrace(Thread.java:1117)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:620)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1832)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1845)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1858)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1929)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1.apply(RDD.scala:920)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1.apply(RDD.scala:918)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:150)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:111)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:316)
	at org.apache.spark.rdd.RDD.foreachPartition(RDD.scala:918)
	at org.apache.spark.sql.DataFrame$$anonfun$foreachPartition$1.apply$mcV$sp(DataFrame.scala:1444)
	at org.apache.spark.sql.DataFrame$$anonfun$foreachPartition$1.apply(DataFrame.scala:1444)
	at org.apache.spark.sql.DataFrame$$anonfun$foreachPartition$1.apply(DataFrame.scala:1444)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:56)
	at org.apache.spark.sql.DataFrame.withNewExecutionId(DataFrame.scala:2086)
	at org.apache.spark.sql.DataFrame.foreachPartition(DataFrame.scala:1443)
	at org.wso2.carbon.analytics.spark.core.sources.AnalyticsRelation.writeDataFrameToDAL(AnalyticsRelation.java:241)
	at org.wso2.carbon.analytics.spark.core.sources.AnalyticsRelation.insert(AnalyticsRelation.java:222)
	at org.apache.spark.sql.execution.datasources.InsertIntoDataSource.run(InsertIntoDataSource.scala:40)
	at org.apache.spark.sql.execution.ExecutedCommand.sideEffectResult$lzycompute(commands.scala:58)
	at org.apache.spark.sql.execution.ExecutedCommand.sideEffectResult(commands.scala:56)
	at org.apache.spark.sql.execution.ExecutedCommand.doExecute(commands.scala:70)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$5.apply(SparkPlan.scala:132)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$5.apply(SparkPlan.scala:130)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:150)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:130)
	at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:55)
	at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:55)
	at org.apache.spark.sql.DataFrame.<init>(DataFrame.scala:145)
	at org.apache.spark.sql.DataFrame.<init>(DataFrame.scala:130)
	at org.apache.spark.sql.DataFrame$.apply(DataFrame.scala:52)
	at org.apache.spark.sql.SQLContext.sql(SQLContext.scala:817)
	at org.wso2.carbon.analytics.spark.core.internal.SparkAnalyticsExecutor.executeQueryLocal(SparkAnalyticsExecutor.java:768)
	... 11 more
Caused by: java.io.FileNotFoundException: /home/asankav/servers/releasetest/ei/alpha/wso2ei-6.1.0-ALPHA/wso2/analytics/tmp/blockmgr-3ba41862-2602-498d-aeb9-50a9918d459f/11/temp_shuffle_c7b89fd2-8842-4421-9ac2-6bd93a656537 (Too many open files)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:286)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:226)
	at org.apache.spark.storage.DiskBlockObjectWriter.open(DiskBlockObjectWriter.scala:88)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:140)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:73)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:41)
	at org.apache.spark.scheduler.Task.run(Task.scala:89)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:227)
	... 3 more
```

**Comment by AnuGayan** [comment]
This was fixed in EI Analytics 6.4.0
